### PR TITLE
[HW] Add `getReferencedModule` to `HWInstanceLike`

### PR DESCRIPTION
**Pull request by nandor**
Added a method to fetch the referenced op of any instance-like ops.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 1]
```diff
@@ -1157,7 +1157,7 @@ void FMemModuleOp::getAsmBlockArgumentNames(
 
 /// Lookup the module or extmodule for the symbol.  This returns null on
 /// invalid IR.
-FModuleLike InstanceOp::getReferencedModule() {
+Operation *InstanceOp::getReferencedModule() {
   auto circuit = (*this)->getParentOfType<CircuitOp>();
   if (!circuit)
     return nullptr;
```
Comment (fabianschuiki) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1163]: I'm wondering if we shouldn't also provide an operation that returns a `FModuleLike` to encode more strongly that FIRRTL instances can only really instantiate `FModuleLike` ops.

File: include/circt/Dialect/HW/HWOpInterfaces.td [hunk 2]
```diff
@@ -138,6 +138,11 @@ def HWInstanceLike : OpInterface<"HWInstanceLike"> {
     "::mlir::StringAttr", "referencedModuleNameAttr", (ins),
     /*methodBody=*/[{}],
     /*defaultImplementation=*/[{ return $_op.getModuleNameAttr().getAttr(); }]>,
+
+    InterfaceMethod<"Get the referenced module",
+    "::mlir::Operation *", "getReferencedModule", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{}]>,
```
Comment (fabianschuiki) [inline include/circt/Dialect/HW/HWOpInterfaces.td:145]: Random thought: does MLIR support making the `HWInstanceLike` interface parametric over the referenced module op type? That could allow us to more strictly enforce that the instantiated ops be e.g. `FModuleLike` for FIRRTL instances.

**Review by fabianschuiki — APPROVED** [review]
This is a nice addition!

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 1 above)
Comment (nandor) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1163]: In the spirit of opaque types, I would keep this as an operation to permit FIRRTL to be composed with other dialects.

File: include/circt/Dialect/HW/HWOpInterfaces.td (hunk 2 above)
Comment (nandor) [reply include/circt/Dialect/HW/HWOpInterfaces.td:145]: Discussed offline: we cannot make this HWModuleLike due to MLIR emission ordering. We also want this to be as generic as possible.

**Pull request merged**